### PR TITLE
Tweak: Added correlation between starting and ending date in Date and Time rule in Popup [ED-9266]

### PR DESCRIPTION
**Pull request by maork-elementor**

### CODE REVIEW

File: assets/dev/js/editor/controls/date-time.js [hunk 1]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
...
+				case '>':
+					if ( startDateTimestamp > endDateTimestamp ) {
+						this.ui.input[ 0 ].value = '';
+					}
+					break;
+				case '<':
+					if ( startDateTimestamp < endDateTimestamp ) {
+						this.ui.input[ 0 ].value = '';
+					}
+					break;
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:52]: maybe you can use `elementor.conditions.compare`:

```JS
// ... 
const startDateTimestamp = new Date( startDate ).getTime();
const endDateTimestamp = new Date( endDate ).getTime();
const operator = '...';

if ( ! elementor.conditions.compare( startDateTimestamp, endDateTimestamp ) ) {
    this.ui.input.val( '' );
}
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/date-time.js (hunk 1 above)
Comment (StyleShit) [reply assets/dev/js/editor/controls/date-time.js:52]: https://github.com/elementor/elementor/blob/dffd09ea02a36d4130dd310c5e2b5b0a4fc615b3/assets/dev/js/editor/utils/conditions.js

File: assets/dev/js/editor/controls/date-time.js [hunk 2]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
 		this.ui.input.flatpickr( options );
 	}
 
+	events() {
+		return {
+			...super.events(),
+			'change @ui.input': 'onBaseInputChange',
+		};
+	}
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:18]: Done

File: assets/dev/js/editor/controls/date-time.js [hunk 3]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
...
+
+	onBaseInputChange() {
+		super.onBaseInputChange( ...arguments );
+		if ( this.model.attributes.datetime_verify ) {
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js]: Done

[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js]: Done

File: assets/dev/js/editor/controls/date-time.js [hunk 4]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
...
+			const startDate = this.options.container.settings.get( this.model.attributes.datetime_verify.control_name );
+			if ( ! startDate ) {
+				return;
+			}
+
+			const endDate = this.ui.input[ 0 ].value;
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:28]: Done

File: assets/dev/js/editor/controls/date-time.js [hunk 5]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
...
+		return {
+			...super.events(),
+			'change @ui.input': 'onBaseInputChange',
+		};
+	}
+
+	onBaseInputChange() {
+		super.onBaseInputChange( ...arguments );
+		if ( this.model.attributes.datetime_verify ) {
+			const startDate = this.options.container.settings.get( this.model.attributes.datetime_verify.control_name );
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js]: Done

File: assets/dev/js/editor/controls/date-time.js [hunk 6]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
...
+			if ( ! startDate ) {
+				return;
+			}
+
+			const endDate = this.ui.input[ 0 ].value;
+			const startDateTimestamp = new Date( startDate ).getTime();
+			const endDateTimestamp = new Date( endDate ).getTime();
+
+			switch ( this.model.attributes.datetime_verify.operator ) {
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js]: Done

File: assets/dev/js/editor/controls/date-time.js [hunk 7]
```diff
@@ -10,6 +10,60 @@ export default class extends ControlBaseDataView {
...
+				case '>=':
+					if ( startDateTimestamp >= endDateTimestamp ) {
+						this.ui.input[ 0 ].value = '';
```
[the remark this answers is not in the document]
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:35]: Done

File: assets/dev/js/editor/controls/date-time.js (hunk 1 above)
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:52]: Nice one, Done.

File: assets/dev/js/editor/controls/date-time.js [hunk 8]
```diff
@@ -10,6 +10,32 @@ export default class extends ControlBaseDataView {
...
 	}
 
+	onBaseInputChange() {
+		super.onBaseInputChange( ...arguments );
+
+		if ( ! this.model.attributes?.validation ) {
+			return;
+		}
+
+		if ( this.model.attributes?.validation.minDate ) {
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:20]: early return

File: assets/dev/js/editor/controls/date-time.js [hunk 9]
```diff
@@ -10,6 +10,32 @@ export default class extends ControlBaseDataView {
...
+			const controlName = this.model.attributes.validation.minDate.control_name;
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:21]:
```JS
const { control_name: controlName, operator } = this.model.attributes.validation.minDate;
```

Maybe? (and then remove the `operator` variable in line 31)

File: assets/dev/js/editor/controls/date-time.js [hunk 10]
```diff
@@ -10,6 +10,32 @@ export default class extends ControlBaseDataView {
...
+			const startDate = this.options.container.settings.get( controlName );
+
+			if ( ! startDate ) {
+				return;
+			}
+
+			const endDate = this.ui.input.val();
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:28]: the current control isn't always the end date, I think we need to add something like `validation.maxDate`

Maybe we can change the API in PHP to indicate that the 2 controls are linked, and then you won't need to specify the validations config there, everything will be validated automatically once the developer linked them. Plus, we probably need to talk with @rami-elementor about it

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/date-time.js (hunk 9 above)
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:21]: Done

File: assets/dev/js/editor/controls/date-time.js (hunk 10 above)
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:28]: @rami-elementor 
What do you think?

File: assets/dev/js/editor/controls/date-time.js (hunk 8 above)
Comment (maork-elementor) [reply assets/dev/js/editor/controls/date-time.js:20]: Done

File: assets/dev/js/editor/controls/date-time.js [hunk 11]
```diff
@@ -10,6 +10,31 @@ export default class extends ControlBaseDataView {
 		this.ui.input.flatpickr( options );
 	}
 
+	onBaseInputChange() {
+		super.onBaseInputChange( ...arguments );
+
+		if ( ! this.model.attributes?.validation ) {
+			return;
+		}
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:18]:
```suggestion
		if ( this.model.attributes?.validation.date_time ) {
			this.validateDateTime();
		}
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/date-time.js [hunk 12]
```diff
@@ -10,6 +10,31 @@ export default class extends ControlBaseDataView {
...
+				return;
+			}
+
+			const endDate = this.ui.input.val();
+			const startDateTimestamp = new Date( startDate ).getTime();
+			const endDateTimestamp = new Date( endDate ).getTime();
+
+			if ( elementor.conditions.compare( startDateTimestamp, endDateTimestamp, operator ) ) {
+				this.ui.input.val( '' );
+			}
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:35]:
```suggestion
		const { control_name: controlName, operator } = this.model.attributes.validation.date_time;
		const startDate = this.options.container.settings.get( controlName );
		
		if ( ! startDate ) {
			return;
		}
		
		const endDate = this.ui.input.val();
		const startDateTimestamp = new Date( startDate ).getTime();
		const endDateTimestamp = new Date( endDate ).getTime();
		
		if ( elementor.conditions.compare( startDateTimestamp, endDateTimestamp, operator ) ) {
			this.ui.input.val( '' );
		}
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/date-time.js [hunk 13]
```diff
@@ -10,6 +10,31 @@ export default class extends ControlBaseDataView {
...
+
+	validateDateTime() {
+		const { control_name: controlName, operator } = this.model.attributes.validation.date_time;
+		const startDate = this.options.container.settings.get( controlName );
+		
+		if ( ! startDate ) {
+			return;
+		}
+		
+		const endDate = this.ui.input.val();
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:29]:
```suggestion
		const startDate = this.options.container.settings.get( controlName );
		const endDate = this.ui.input.val();
		
		if ( ! startDate || ! endDate ) {
			return;
		}
		
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/date-time.js [hunk 14]
```diff
@@ -10,6 +10,31 @@ export default class extends ControlBaseDataView {
 		this.ui.input.flatpickr( options );
 	}
 
+	onBaseInputChange() {
+		super.onBaseInputChange( ...arguments );
+
+		if ( this.model.attributes?.validation.date_time ) {
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:16]:
```suggestion
		if ( this.model.get( 'validation' )?.date_time ) {
```

File: assets/dev/js/editor/controls/date-time.js [hunk 15]
```diff
@@ -10,6 +10,31 @@ export default class extends ControlBaseDataView {
...
+			this.validateDateTime();
+		}
+	}
+
+	validateDateTime() {
+		const { control_name: controlName, operator } = this.model.attributes.validation.date_time;
```
Comment (StyleShit) [inline assets/dev/js/editor/controls/date-time.js:22]:
```suggestion
		const { control_name: controlName, operator } = this.model.get( 'validation' ).date_time;
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

**Review by StyleShit — APPROVED** [review]

**Review by rotemee — APPROVED** [review]

**Pull request merged**